src/logging.c

```diff
@@ -24,81 +24,271 @@ extern "C"
-    if (NULL == ret && strcmp(output_format, "") != 0) {
+    const char * ret_str = rcutils_get_env("RCUTILS_CONSOLE_OUTPUT_FORMAT", &output_format);
+    if (NULL == ret_str && strcmp(output_format, "") != 0) {
       size_t chars_to_copy = strlen(output_format);
       if (chars_to_copy > RCUTILS_LOGGING_MAX_OUTPUT_FORMAT_LEN - 1) {
         chars_to_copy = RCUTILS_LOGGING_MAX_OUTPUT_FORMAT_LEN - 1;
       }
       memcpy(g_rcutils_logging_output_format_string, output_format, chars_to_copy);
       g_rcutils_logging_output_format_string[chars_to_copy] = '\0';
     } else {
-      if (NULL != ret) {
+      if (NULL != ret_str) {
         fprintf(
           stderr,
           "Failed to get output format from env. variable: %s. Using default output format.\n",
-          ret);
+          ret_str);
+        ret = RCUTILS_RET_INVALID_ARGUMENT;
       }
       memcpy(g_rcutils_logging_output_format_string, g_rcutils_logging_default_output_format,
         strlen(g_rcutils_logging_default_output_format) + 1);
     }
+
+    if (!rcutils_allocator_is_valid(&allocator)) {
+      fprintf(
+        stderr,
+        "Provided allocator is invalid. Using the default allocator.\n");
+      ret = RCUTILS_RET_INVALID_ARGUMENT;
+      allocator = rcutils_get_default_allocator();
+    }
+    g_rcutils_logging_allocator = allocator;
+    g_rcutils_logging_severities_map = rcutils_get_zero_initialized_string_map();
+    rcutils_ret_t string_map_ret = rcutils_string_map_init(
+      &g_rcutils_logging_severities_map, 0, g_rcutils_logging_allocator);
+    if (string_map_ret != RCUTILS_RET_OK) {
+      fprintf(
+        stderr,
+        "Failed to initialize map for logger severities. Severities will not be configurable.\n");
+      g_rcutils_logging_severities_map_valid = false;
+      ret = RCUTILS_RET_STRING_MAP_INVALID;
+    } else {
+      g_rcutils_logging_severities_map_valid = true;
+    }
+
     g_rcutils_logging_initialized = true;
   }
+  return ret;
+}
+
+rcutils_ret_t rcutils_logging_shutdown()
+{
+  if (!g_rcutils_logging_initialized) {
+    return RCUTILS_RET_OK;
+  }
+  rcutils_ret_t ret = RCUTILS_RET_OK;
+  if (g_rcutils_logging_severities_map_valid) {
+    rcutils_ret_t ret = rcutils_string_map_fini(&g_rcutils_logging_severities_map);
+    if (ret != RCUTILS_RET_OK) {
+      fprintf(stderr, "Failed to finalize logging severities map: return code %d", ret);
+      ret = RCUTILS_RET_LOGGING_SEVERITY_MAP_INVALID;
+    }
+    g_rcutils_logging_severities_map_valid = false;
+  }
+  g_rcutils_logging_initialized = false;
+  return ret;
 }
 
 rcutils_logging_output_handler_t rcutils_logging_get_output_handler()
 {
+  RCUTILS_LOGGING_AUTOINIT
   return g_rcutils_logging_output_handler;
 }
 
 void rcutils_logging_set_output_handler(rcutils_logging_output_handler_t function)
 {
-  g_rcutils_logging_output_handler = function;
+  RCUTILS_LOGGING_AUTOINIT
+    g_rcutils_logging_output_handler = function;
```

Comment:
Indentation seems to be off?

Same below.

Reply:
yeah this is what uncrustify wanted: I'll disable what's making it complain

Reply:
a78c8862ff06e7d75a5eac9c4cb6f92eb92d89cb